src/util/listener-collection.ts

```diff
@@ -0,0 +1,75 @@
+export type Listener = {
+  eventTarget: EventTarget;
+  eventType: string;
+  listener: EventListener;
+  options?: AddEventListenerOptions;
+};
+
+/**
+ * Return the event type that a listener will receive.
+ *
+ * For example `EventType<HTMLElement, 'keydown'>` evaluates to `KeyboardEvent`.
+ *
+ * The event type is extracted from the target's `on${Type}` property (eg.
+ * `HTMLElement.onkeydown` here) If there is no such property, the type defaults
+ * to `Event`.
+ */
+type EventType<
+  Target extends EventTarget,
+  TypeName extends string
+> = `on${TypeName}` extends keyof Target
+  ? Target[`on${TypeName}`] extends ((...args: any[]) => void) | null
+    ? Parameters<NonNullable<Target[`on${TypeName}`]>>[0]
+    : Event
+  : Event;
+
+/**
+ * Utility that provides a way to conveniently remove a set of DOM event
+ * listeners when they are no longer needed.
+ */
+export class ListenerCollection {
+  private _listeners: Map<symbol, Listener>;
+
+  constructor() {
+    this._listeners = new Map();
+  }
+
+  /**
+   * Add a listener and return an ID that can be used to remove it later
+   */
+  add<ListenerTarget extends EventTarget, Type extends string>(
+    eventTarget: ListenerTarget,
+    eventType: Type,
+    listener: (event: EventType<ListenerTarget, Type>) => void,
+    options?: AddEventListenerOptions
+  ) {
```

Comment:
It might be helpful for readers to indicate explicitly that the return type is a `symbol`.